### PR TITLE
[Common/Subplugin] Internal API to handle subplugins  - @open sesame 12/01 12:42

### PR DESCRIPTION
**Pull request by myungjoo**
Any subplugin-using elements may call get_subplugin () to
get internal data (set of callbacks) of the subplugin.
    
Next Step: modify tensor_decoder to use nnstreamer_subplugin.h
and test with it.
    
Signed-off-by: MyungJoo Ham <myungjoo.ham@samsung.com>

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [ ]Passed [ ]Failed [x]Skipped

### CODE REVIEW

File: gst/nnstreamer/nnstreamer_subplugin.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * @return TRUE if regitered. FALSE if rejected (duplicated) or error.
+ */
+extern gboolean
+register_subplugin (subpluginType type, const char *name, const void *data);
+
+/**
+ * @brief Unregister the subplugin.
+ * @param[in] type Subplugin type
+ * @param[in] name Subplugin Name. The filename should be subplugin_prefixes[type]${name}.so
+ * @param[in] data The registered data
```
Comment (jaeyun-jung) [inline gst/nnstreamer/nnstreamer_subplugin.h]: remove this line (data)

File: gst/nnstreamer/nnstreamer_subplugin.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    if (g_strcmp0 (nsdata->name, name)) {
+      GST_ERROR ("nnstreamer_subplugin of %s (%s) is broken (name mismatch)",
+          name, fullpath);
+      dlclose (handle);
+      goto error;
+    }
+
+    g_assert (TRUE == register_subplugin (type, name, data));
```
Comment (jaeyun-jung) [inline gst/nnstreamer/nnstreamer_subplugin.c]: register_subplugin (type, name, data)
at this point, data is null, is it correct?

File: gst/nnstreamer/nnstreamer_subplugin.c [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  if (subplugins[type] == NULL)
+    subplugins[type] = g_hash_table_new (_sphash, _strequal);
+
+  G_LOCK (splock);
+  if (g_hash_table_contains (subplugins[type], spdata))
+    ret = FALSE;
+
+  if (ret == TRUE)
+    g_hash_table_add (subplugins[type], spdata);
+
```
Comment (jijoongmoon) [inline gst/nnstreamer/nnstreamer_subplugin.c]: I'm wondering g_hash_table_add() is always successful? I think if it does not, then we may need free spdata.

**Comment by myungjoo** [comment]
Fixing up more. Please wait.

File: gst/nnstreamer/nnstreamer_subplugin.c (hunk 2 above)
Comment (myungjoo) [reply gst/nnstreamer/nnstreamer_subplugin.c]: Fixed. Thanks.

File: gst/nnstreamer/nnstreamer_subplugin.c (hunk 3 above)
Comment (myungjoo) [reply gst/nnstreamer/nnstreamer_subplugin.c]: It's changed to g_hash_table_insert(), which does not "FAIL-RETURN".
It returns FALSE if the key already exists; however, it simply overwrites if so.
Thus, we do not need to worry about spdata. We just need to worry about the overwritten old data.

Comment (myungjoo) [reply gst/nnstreamer/nnstreamer_subplugin.c]: And the old data is now handled by _spdata_destroy

**Pull request merged**